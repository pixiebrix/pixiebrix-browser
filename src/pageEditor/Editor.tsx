/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./Editor.module.scss";

import React, { useCallback, useContext, useEffect, useMemo } from "react";
import Sidebar from "@/pageEditor/sidebar/Sidebar";
import { useDispatch, useSelector } from "react-redux";
import { PageEditorTabContext } from "@/pageEditor/context";
import { selectExtensions } from "@/store/extensionsSelectors";
import PermissionsPane from "@/pageEditor/panes/PermissionsPane";
import BetaPane from "@/pageEditor/panes/BetaPane";
import InsertMenuItemPane from "@/pageEditor/panes/insert/InsertMenuItemPane";
import InsertPanelPane from "@/pageEditor/panes/insert/InsertPanelPane";
import NoExtensionSelectedPane from "@/pageEditor/panes/NoExtensionSelectedPane";
import NoExtensionsPane from "@/pageEditor/panes/NoExtensionsPane";
import WelcomePane from "@/pageEditor/panes/WelcomePane";
import EditorPane from "@/pageEditor/panes/EditorPane";
import useInstallState from "@/pageEditor/hooks/useInstallState";
import useEscapeHandler from "@/pageEditor/hooks/useEscapeHandler";
import GenericInsertPane from "@/pageEditor/panes/insert/GenericInsertPane";
import { ADAPTERS } from "@/pageEditor/extensionPoints/adapter";
import { actions } from "@/pageEditor/slices/editorSlice";
import { useGetMarketplaceListingsQuery } from "@/services/api";
import { cancelSelect } from "@/contentScript/messenger/api";
import { thisTab } from "@/pageEditor/utils";
import {
  selectIsAddToRecipeModalVisible,
  selectIsCreateRecipeModalVisible,
  selectIsRemoveFromRecipeModalVisible,
  selectIsSaveAsNewRecipeModalVisible,
} from "@/pageEditor/slices/editorSelectors";
import RecipePane from "@/pageEditor/panes/RecipePane";
import { selectSessionId } from "@/pageEditor/slices/sessionSelectors";
import { reportEvent } from "@/telemetry/events";
import { RootState } from "@/pageEditor/pageEditorTypes";
import AddToRecipeModal from "@/pageEditor/sidebar/AddToRecipeModal";
import RemoveFromRecipeModal from "@/pageEditor/sidebar/RemoveFromRecipeModal";
import CreateRecipeModal from "@/pageEditor/sidebar/CreateRecipeModal";
import SaveAsNewRecipeModal from "@/pageEditor/sidebar/SaveAsNewRecipeModal";
import useFlags from "@/hooks/useFlags";
import RestrictedPane from "@/pageEditor/panes/RestrictedPane";

const selectEditor = ({ editor }: RootState) => editor;

const Editor: React.FunctionComponent = () => {
  const { tabState, connecting } = useContext(PageEditorTabContext);
  const installed = useSelector(selectExtensions);
  const dispatch = useDispatch();
  const { restrict } = useFlags();

  const sessionId = useSelector(selectSessionId);
  useEffect(() => {
    reportEvent("PageEditorSessionStart", {
      sessionId,
    });

    return () => {
      reportEvent("PageEditorSessionEnd", {
        sessionId,
      });
    };
  }, [sessionId]);

  // Async fetch marketplace content to the Redux so it's pre-fetched for rendering in the Brick Selection modal
  useGetMarketplaceListingsQuery();

  const {
    inserting,
    elements,
    activeElementId,
    activeRecipeId,
    error: editorError,
    beta,
  } = useSelector(selectEditor);

  const cancelInsert = useCallback(async () => {
    dispatch(actions.toggleInsert(null));
    await cancelSelect(thisTab);
  }, [dispatch]);

  useEscapeHandler(cancelInsert, inserting != null);

  const { availableDynamicIds, unavailableCount, loading } = useInstallState(
    installed,
    elements
  );

  const isAddToRecipeModalVisible = useSelector(
    selectIsAddToRecipeModalVisible
  );
  const isRemoveFromRecipeModalVisible = useSelector(
    selectIsRemoveFromRecipeModalVisible
  );
  const isSaveAsNewRecipeModalVisible = useSelector(
    selectIsSaveAsNewRecipeModalVisible
  );
  const isCreateRecipeModalVisible = useSelector(
    selectIsCreateRecipeModalVisible
  );

  const body = useMemo(() => {
    if (restrict("page-editor")) {
      return <RestrictedPane />;
    }

    // Need to explicitly check for `false` because hasPermissions will be undefined if pending/error
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-boolean-literal-compare
    if (tabState.hasPermissions === false && !connecting) {
      // Check `connecting` to optimistically show the main interface while the devtools are connecting to the page.
      return <PermissionsPane />;
    }

    if (editorError && beta) {
      return <BetaPane />;
    }

    if (inserting) {
      switch (inserting) {
        case "menuItem":
          return <InsertMenuItemPane cancel={cancelInsert} />;
        case "panel":
          return <InsertPanelPane cancel={cancelInsert} />;
        default:
          return (
            <GenericInsertPane
              cancel={cancelInsert}
              config={ADAPTERS.get(inserting)}
            />
          );
      }
    } else if (editorError) {
      return (
        <div className="p-2">
          <span className="text-danger">{editorError}</span>
        </div>
      );
    } else if (activeElementId) {
      return <EditorPane />;
    } else if (activeRecipeId) {
      return <RecipePane />;
    } else if (connecting || loading) {
      // Don't show anything while it's loading
      return null;
    } else if (
      availableDynamicIds?.size ||
      installed.length > unavailableCount
    ) {
      return <NoExtensionSelectedPane />;
    } else if (installed.length > 0) {
      return <NoExtensionsPane unavailableCount={unavailableCount} />;
    } else {
      return <WelcomePane />;
    }
  }, [
    restrict,
    loading,
    tabState.hasPermissions,
    connecting,
    editorError,
    beta,
    inserting,
    activeElementId,
    activeRecipeId,
    availableDynamicIds?.size,
    installed.length,
    unavailableCount,
    cancelInsert,
  ]);

  return (
    <>
      <div className={styles.root}>
        {!(inserting || restrict("page-editor")) && <Sidebar />}
        {body}
      </div>

      {isAddToRecipeModalVisible && <AddToRecipeModal />}

      {isRemoveFromRecipeModalVisible && <RemoveFromRecipeModal />}

      {isSaveAsNewRecipeModalVisible && <SaveAsNewRecipeModal />}

      {isCreateRecipeModalVisible && <CreateRecipeModal />}
    </>
  );
};

export default Editor;
