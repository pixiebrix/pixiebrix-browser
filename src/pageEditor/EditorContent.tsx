/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useEffect } from "react";
import { useSelector } from "react-redux";
import { selectSessionId } from "@/pageEditor/slices/sessionSelectors";
import reportEvent from "@/telemetry/reportEvent";
import { Events } from "@/telemetry/events";
import { useGetMarketplaceListingsQuery } from "@/services/api";
import NoTabAccessPane from "@/pageEditor/panes/NoTabAccessPane";
import BetaPane from "@/pageEditor/panes/BetaPane";
import EditorPane from "@/pageEditor/panes/EditorPane";
import RecipePane from "@/pageEditor/panes/RecipePane";
import NoExtensionSelectedPane from "@/pageEditor/panes/NoExtensionSelectedPane";
import NoExtensionsPane from "@/pageEditor/panes/NoExtensionsPane";
import WelcomePane from "@/pageEditor/panes/WelcomePane";
import {
  selectActiveElementId,
  selectActiveRecipeId,
  selectErrorState,
  selectExtensionAvailability,
  selectNotDeletedExtensions,
} from "@/pageEditor/slices/editorSelectors";
import {
  selectTabHasPermissions,
  selectTabIsConnectingToContentScript,
} from "@/pageEditor/tabState/tabStateSelectors";
import useCurrentUrl from "@/pageEditor/hooks/useCurrentUrl";
import { getErrorMessage } from "@/errors/errorHelpers";
import Alert from "@/components/Alert";
import styles from "./EditorContent.module.scss";

const EditorContent: React.FC = () => {
  const tabHasPermissions = useSelector(selectTabHasPermissions);
  const isConnectingToContentScript = useSelector(
    selectTabIsConnectingToContentScript
  );
  const installed = useSelector(selectNotDeletedExtensions);
  const sessionId = useSelector(selectSessionId);
  const { isBetaError, editorError } = useSelector(selectErrorState);
  const activeElementId = useSelector(selectActiveElementId);
  const activeRecipeId = useSelector(selectActiveRecipeId);
  const {
    availableDynamicIds,
    unavailableDynamicCount,
    unavailableInstalledCount,
    isPendingInstalledExtensions,
    isPendingDynamicExtensions,
  } = useSelector(selectExtensionAvailability);

  const url = useCurrentUrl();

  useEffect(() => {
    console.debug("EditorContent debug effect", {
      url,
      isPendingInstalledExtensions,
      isPendingDynamicExtensions,
      isConnectingToContentScript,
    });
  }, [
    url,
    isPendingInstalledExtensions,
    isPendingDynamicExtensions,
    isConnectingToContentScript,
  ]);

  const unavailableCount = unavailableInstalledCount + unavailableDynamicCount;
  const isPendingExtensions =
    isPendingInstalledExtensions || isPendingDynamicExtensions;

  // Fetch-and-cache marketplace content for rendering in the Brick Selection modal
  useGetMarketplaceListingsQuery();

  useEffect(() => {
    reportEvent(Events.PAGE_EDITOR_SESSION_START, {
      sessionId,
    });

    return () => {
      reportEvent(Events.PAGE_EDITOR_SESSION_END, {
        sessionId,
      });
    };
  }, [sessionId]);

  // Always show the main error if present - keep this first
  if (editorError) {
    return (
      <div className={styles.alertContainer}>
        <Alert variant="danger">{getErrorMessage(editorError)}</Alert>
      </div>
    );
  }

  if (!tabHasPermissions && !isConnectingToContentScript) {
    // Check `connecting` to optimistically show the main interface while the devtools are connecting to the page.
    return <NoTabAccessPane />;
  }

  // Show generic error for beta features
  if (isBetaError) {
    return <BetaPane />;
  }

  if (activeElementId) {
    return <EditorPane />;
  }

  if (activeRecipeId) {
    return <RecipePane />;
  }

  if (isPendingExtensions || isConnectingToContentScript) {
    // Avoid flashing the panes below while the state is loading. This condition should probably
    // not be moved below <NoExtensionSelectedPane>, <NoExtensionsPane>, or <WelcomePane>.
    // It loads fast enough to not require a <Loader> either.
    // https://github.com/pixiebrix/pixiebrix-extension/pull/3611
    return null;
  }

  if (availableDynamicIds?.length > 0 || installed.length > unavailableCount) {
    return <NoExtensionSelectedPane />;
  }

  if (installed.length > 0) {
    return <NoExtensionsPane unavailableCount={unavailableCount} />;
  }

  return <WelcomePane />;
};

export default EditorContent;
