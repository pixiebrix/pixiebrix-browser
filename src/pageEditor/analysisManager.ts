/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import BlockTypeAnalysis from "@/analysis/analysisVisitors/blockTypeAnalysis";
import ExtensionUrlPatternAnalysis from "@/analysis/analysisVisitors/extensionUrlPatternAnalysis";
import OutputKeyAnalysis from "@/analysis/analysisVisitors/outputKeyAnalysis";
import RenderersAnalysis from "@/analysis/analysisVisitors/renderersAnalysis";
import TemplateAnalysis from "@/analysis/analysisVisitors/templateAnalysis";
import TraceAnalysis from "@/analysis/analysisVisitors/traceAnalysis";
import ReduxAnalysisManager from "@/analysis/ReduxAnalysisManager";
import { type UUID } from "@/core";
import { type TraceRecord } from "@/telemetry/trace";
import { isAnyOf, type PayloadAction } from "@reduxjs/toolkit";
import { type RootState } from "./pageEditorTypes";
import { actions as editorActions } from "@/pageEditor/slices/editorSlice";
import runtimeSlice from "./slices/runtimeSlice";
import RequestPermissionAnalysis from "@/analysis/analysisVisitors/requestPermissionAnalysis";
import FormBrickAnalysis from "@/analysis/analysisVisitors/formBrickAnalysis";
import { selectActiveElementTraces } from "./slices/runtimeSelectors";
import VarAnalysis from "@/analysis/analysisVisitors/varAnalysis/varAnalysis";
import analysisSlice from "@/analysis/analysisSlice";
import RegexAnalysis from "@/analysis/analysisVisitors/regexAnalysis";

const runtimeActions = runtimeSlice.actions;

const pageEditorAnalysisManager = new ReduxAnalysisManager();

// These actions will be used with every analysis so the annotation path is up-to-date
// with the node position in the pipeline
const nodeListMutationActions = [
  editorActions.addNode,
  editorActions.moveNode,
  editorActions.removeNode,
] as const;

// The order in which the analysis are registered is important.
// The first analysis registered will be the first to run.
// When multiple actions (e.g. typing) trigger analysis, the later analysis have more chances to get aborted.
// Try to put the faster analysis first, and the slower ones at the end.

pageEditorAnalysisManager.registerAnalysisEffect(
  (
    action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
    state: RootState
  ) => {
    // TraceAnalysis filter the trace errors, thus
    // selecting all records here to avoid double filtering
    const records = selectActiveElementTraces(state);

    return new TraceAnalysis(records);
  },
  {
    matcher: isAnyOf(
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new BlockTypeAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new FormBrickAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RenderersAnalysis(),
  {
    matcher: isAnyOf(...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new TemplateAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new ExtensionUrlPatternAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(
  () => new RequestPermissionAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

pageEditorAnalysisManager.registerAnalysisEffect(() => new RegexAnalysis(), {
  matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
});

const varAnalysisFactory = (
  action: PayloadAction<{ extensionId: UUID; records: TraceRecord[] }>,
  state: RootState
) => {
  const records = selectActiveElementTraces(state);
  return new VarAnalysis(records);
};

// OutputKeyAnalysis seems to be the slowest one, so we register it in the end
pageEditorAnalysisManager.registerAnalysisEffect(
  () => new OutputKeyAnalysis(),
  {
    matcher: isAnyOf(editorActions.editElement, ...nodeListMutationActions),
  }
);

// VarAnalysis is not the slowest itself, but it triggers a post-analysis action,
// so it is the last one
pageEditorAnalysisManager.registerAnalysisEffect(
  varAnalysisFactory,
  {
    matcher: isAnyOf(
      editorActions.editElement,
      runtimeActions.setExtensionTrace,
      ...nodeListMutationActions
    ),
  },
  {
    postAnalysisAction(analysis, extensionId, listenerApi) {
      listenerApi.dispatch(
        analysisSlice.actions.setKnownVars({
          extensionId,
          vars: analysis.getKnownVars(),
        })
      );
    },
  }
);

export default pageEditorAnalysisManager;
