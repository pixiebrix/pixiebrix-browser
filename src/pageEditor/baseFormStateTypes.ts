/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Except } from "type-fest";
import {
  type IntegrationDependency,
  type IntegrationDependencyV1,
  type IntegrationDependencyV2,
} from "@/types/integrationTypes";
import { type ApiVersion, type OptionsArgs } from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type StarterBrickType } from "@/starterBricks/types";
import { type Permissions } from "webextension-polyfill";
import { type ModComponentBase } from "@/types/modComponentTypes";
import { type ModOptionsDefinition } from "@/types/modDefinitionTypes";
import {
  type BrickPipeline,
  type NormalizedAvailability,
} from "@/bricks/types";
import { type Metadata, type RegistryId } from "@/types/registryTypes";

/**
 * A simplified type for ReaderConfig to prevent TypeScript reporting problems with infinite type instantiation
 * @see ReaderConfig
 */
export type SingleLayerReaderConfig =
  | RegistryId
  | RegistryId[]
  | Record<string, RegistryId>;

export type BaseExtensionPointState = {
  metadata: Metadata;
  definition: {
    type: StarterBrickType;
    // We're currently not allowing users to modify readers in the page editor
    reader: SingleLayerReaderConfig;
    isAvailable: NormalizedAvailability;
  };
};

export interface BaseExtensionState {
  blockPipeline: BrickPipeline;
}

/**
 * @deprecated - Do not use versioned state types directly
 */
export interface BaseFormStateV1<
  TExtension extends BaseExtensionState = BaseExtensionState,
  TExtensionPoint extends BaseExtensionPointState = BaseExtensionPointState
> {
  /**
   * The apiVersion of the brick definition, controlling how PixieBrix interprets brick definitions
   * @see ApiVersion
   */
  readonly apiVersion: ApiVersion;

  /**
   * The extension uuid
   */
  readonly uuid: UUID;

  /**
   * The type of the extensionPoint
   */
  readonly type: StarterBrickType;

  /**
   * True if the extensionPoint exists in the registry
   */
  installed?: boolean;

  /**
   * True if the extension should be allowed to auto-reload. In general, only extensions that require user
   * interaction to trigger should be allowed to auto-reload. Otherwise, PixieBrix might end up spamming an API
   */
  autoReload?: boolean;

  /**
   * User-provided name to identify the extension
   */
  label: string;

  /**
   * The input options from the extension's blueprint
   * @since 1.4.3
   */
  optionsArgs: OptionsArgs;

  services: IntegrationDependencyV1[];

  /**
   * The extra permissions required by the extension
   * @since 1.7.0
   */
  permissions: Permissions.Permissions;

  extensionPoint: TExtensionPoint;

  extension: TExtension;

  /**
   * Information about the recipe (i.e., blueprint) used to install the extension, or `undefined` if the extension
   * is not part of a recipe.
   * @see ModComponentBase._recipe
   */
  recipe: ModComponentBase["_recipe"] | undefined;

  /**
   * Information about the recipe (i.e., blueprint) options,
   * or `undefined` if the extension is not part of a recipe.
   * @see ModDefinition.options
   */
  optionsDefinition?: ModOptionsDefinition;
}

/**
 * @deprecated - Do not use versioned state types directly
 */
export type BaseFormStateV2<
  TExtension extends BaseExtensionState = BaseExtensionState,
  TExtensionPoint extends BaseExtensionPointState = BaseExtensionPointState
> = Except<BaseFormStateV1<TExtension, TExtensionPoint>, "services"> & {
  /**
   * The integration dependencies configured for the extension
   *
   * @since 1.7.41 renamed from `services` to `integrationDependencies`, also
   * changed from IntegrationDependencyV1 to IntegrationDependencyV2
   */
  integrationDependencies: IntegrationDependencyV2[];
};
export type BaseFormState<
  TExtension extends BaseExtensionState = BaseExtensionState,
  TExtensionPoint extends BaseExtensionPointState = BaseExtensionPointState
> = Except<
  BaseFormStateV2<TExtension, TExtensionPoint>,
  "integrationDependencies"
> & {
  /**
   * Using the un-versioned type
   */
  integrationDependencies: IntegrationDependency[];
};
