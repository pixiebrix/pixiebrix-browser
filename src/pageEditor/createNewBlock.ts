/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { BlockConfig } from "@/blocks/types";
import { defaultBlockConfig } from "@/blocks/util";
import { RegistryId, Schema } from "@/core";
import { uuidv4 } from "@/types/helpers";
import { getExampleBlockConfig } from "./exampleBlockConfigs";

export function createNewBlock(
  blockId: RegistryId,
  blockInputSchema?: Schema
): BlockConfig {
  return {
    id: blockId,
    instanceId: uuidv4(),
    config:
      getExampleBlockConfig(blockId) ??
      (blockInputSchema == null ? {} : defaultBlockConfig(blockInputSchema)),
  };
}
