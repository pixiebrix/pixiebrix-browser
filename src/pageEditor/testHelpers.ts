/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import {
  createRenderHookWithWrappers,
  createRenderWithWrappers,
} from "@/testUtils/testHelpers";
import analysisSlice from "@/analysis/analysisSlice";
import pageEditorAnalysisManager from "./analysisManager";
import { tabStateSlice } from "@/pageEditor/tabState/tabStateSlice";
import { appApi } from "@/services/api";
import { modDefinitionsSlice } from "@/modDefinitions/modDefinitionsSlice";
import { modDefinitionsMiddleware } from "@/modDefinitions/modDefinitionsListenerMiddleware";

const configureStoreForTests = () =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      analysis: analysisSlice.reducer,
      tabState: tabStateSlice.reducer,
      modDefinitions: modDefinitionsSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
    middleware(getDefaultMiddleware) {
      /* eslint-disable unicorn/prefer-spread -- It's not Array#concat, can't use spread */
      return getDefaultMiddleware()
        .concat(appApi.middleware)
        .concat(pageEditorAnalysisManager.middleware)
        .concat(modDefinitionsMiddleware);
      /* eslint-enable unicorn/prefer-spread */
    },
  });

const renderWithWrappers = createRenderWithWrappers(configureStoreForTests);
const renderHookWithWrappers = createRenderHookWithWrappers(
  configureStoreForTests
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
export { renderHookWithWrappers as renderHook };
