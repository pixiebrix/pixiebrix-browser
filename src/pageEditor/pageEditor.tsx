/*
 * Copyright (C) 2024 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import "@/vendors/bootstrapWithoutRem.css";
import "@/utils/global.scss";
import "@/utils/layout.scss";

import "@/extensionContext";
import "@/development/darkMode.js";

import { initMessengerLogging } from "@/development/messengerLogging";

import ReactDOM from "react-dom";
import React from "react";
import Panel from "@/pageEditor/Panel";
import { watchNavigation } from "@/pageEditor/protocol";
import { initToaster } from "@/utils/notify";
import { initRuntimeLogging } from "@/development/runtimeLogging";
import { initPerformanceMonitoring } from "@/telemetry/performance";
import { setPlatform } from "@/platform/platformContext";
import extensionPagePlatform from "@/extensionPages/extensionPagePlatform";

async function init() {
  setPlatform(extensionPagePlatform);
  void initMessengerLogging();
  void initRuntimeLogging();
  try {
    await initPerformanceMonitoring();
  } catch (error) {
    console.error("Failed to initialize performance monitoring", error);
  }

  watchNavigation();
  initToaster();

  ReactDOM.render(<Panel />, document.querySelector("#container"));
}

void init();
