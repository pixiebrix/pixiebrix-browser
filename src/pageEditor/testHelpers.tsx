/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { configureStore } from "@reduxjs/toolkit";
import { authSlice } from "@/auth/authSlice";
import extensionsSlice from "@/store/extensionsSlice";
import servicesSlice from "@/store/servicesSlice";
import settingsSlice from "@/store/settingsSlice";
import { editorSlice } from "@/pageEditor/slices/editorSlice";
import sessionSlice from "@/pageEditor/slices/sessionSlice";
import { savingExtensionSlice } from "@/pageEditor/panes/save/savingExtensionSlice";
import runtimeSlice from "@/pageEditor/slices/runtimeSlice";
import { logSlice } from "@/components/logViewer/logSlice";
import { createRenderWithWrappers } from "@/testUtils/testHelpers";
import { appApi } from "@/services/api";

const renderWithWrappers = createRenderWithWrappers(() =>
  configureStore({
    reducer: {
      auth: authSlice.reducer,
      options: extensionsSlice.reducer,
      services: servicesSlice.reducer,
      settings: settingsSlice.reducer,
      editor: editorSlice.reducer,
      session: sessionSlice.reducer,
      savingExtension: savingExtensionSlice.reducer,
      runtime: runtimeSlice.reducer,
      logs: logSlice.reducer,
      [appApi.reducerPath]: appApi.reducer,
    },
  })
);

// eslint-disable-next-line import/export -- re-export RTL
export * from "@testing-library/react";
// eslint-disable-next-line import/export -- override render
export { renderWithWrappers as render };
