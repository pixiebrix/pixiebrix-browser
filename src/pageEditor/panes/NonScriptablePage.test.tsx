/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import NonScriptablePage from "@/pageEditor/NonScriptablePage";
import { render, screen } from "@/pageEditor/testHelpers";
import { waitFor } from "@testing-library/react";

describe("NonScriptablePage", () => {
  test("it renders", () => {
    const rendered = render(<NonScriptablePage url="https://test.url" />);
    expect(rendered.asFragment()).toMatchSnapshot();
  });

  test("it renders right copy when the URL is HTTP", async () => {
    render(<NonScriptablePage url="http://example.com" />);

    await waitFor(() => {
      expect(
        screen.getByText("PixieBrix cannot modify insecure HTTP pages")
      ).not.toBeNull();
    });
  });
});
