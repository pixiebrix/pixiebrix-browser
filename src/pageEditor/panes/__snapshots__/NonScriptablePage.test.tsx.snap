// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NonScriptablePage it renders 1`] = `
<DocumentFragment>
  <div
    class="root container-fluid"
  >
    <div
      class="my-auto"
    >
      <div
        class="paneRow row"
      >
        <div
          class="col-lg-9"
        >
          <h4
            class="callout"
          >
            Get started with PixieBrix
          </h4>
          <p>
            This is the PixieBrix Page Editor. Use the Page Editor to create web page modifications, called mods.
          </p>
          <p>
            To start, navigate to a web page that you'd like to modify and then click the "Add" button.
          </p>
          <p>
            Try the 
            <a
              href="https://www.pixiebrix.com/playground/playground"
              rel="noopener noreferrer"
              target="_blank"
            >
              PixieBrix Playground
            </a>
            .
          </p>
        </div>
        <div
          class="d-flex align-items-center justify-content-center col-lg-3"
        >
          <img
            alt=""
            class="illustration"
            src="test-file-stub"
            width="100%"
          />
        </div>
      </div>
      <div
        class="paneRowWithDivider row"
      >
        <div
          class="col-lg-9"
        >
          <div>
            <h4
              class="tinyCallout"
            >
              Need more help?
            </h4>
            <p>
              Visit the 
              <a
                href="https://docs.pixiebrix.com/quick-start-guide"
                rel="noopener noreferrer"
                target="_blank"
              >
                Quick Start Guide
              </a>
               or ask questions in the 
              <a
                href="https://pixiebrixcommunity.slack.com/join/shared_invite/zt-13gmwdijb-Q5nVsSx5wRLmRwL3~lsDww#/shared-invite/email"
                rel="noopener noreferrer"
                target="_blank"
              >
                Slack Community
              </a>
              . 
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
